ambry-store/src/main/java/com/github/ambry/store/BlobStoreCompactor.java

```diff
@@ -1295,4 +998,533 @@ private void endCompaction() {
+        // the PUT is going away in this cycle, it is safe to remove the TTL update also)
+        //
+        // For condition one, we will keep TTL_UPDATE, for condition 2, we will remove it.
+        if (isOffsetUnderCompaction(srcValue.getOffset())) {
+          logger.trace(
+              "TTL update of {} in segment with start offset {} in {} is not valid because the corresponding PUT entry"
+                  + " {} will be compacted in this cycle ({} are being compacted in this cycle)", key,
+              indexSegmentStartOffset, storeId, srcValue,
+              compactionLog.getCompactionDetails().getLogSegmentsUnderCompaction());
+        } else {
+          // PUT entry exists in both the src and tgt,
+          // OR PUT entry exists in source and the offset of the source entry is not under compaction in this cycle
+          // therefore this TTL update entry cannot be compacted
+          valid = true;
+        }
+      }
+      return valid;
+    }
+
+    @Override
+    public boolean alreadyExists(PersistentIndex idx, FileSpan searchSpan, StoreKey key, IndexValue srcValue)
+        throws StoreException {
+      IndexValue value = idx.findKey(key, searchSpan, EnumSet.allOf(PersistentIndex.IndexEntryType.class));
+      if (value == null) {
+        return false;
+      }
+      if (value.getLifeVersion() > srcValue.getLifeVersion()) {
+        // If the IndexValue in the previous index has a lifeVersion higher than source value, then this
+        // must be a duplicate.
+        return true;
+      } else if (value.getLifeVersion() < srcValue.getLifeVersion()) {
+        // If the IndexValue in the previous index has a lifeVersion lower than source value, then this
+        // is not a duplicate.
+        return false;
+      } else {
+        // When the lifeVersions are the same, the order of the record are **FIX** as
+        // P/U -> T -> D
+        if (value.isDelete()) {
+          return true;
+        } else if (value.isTtlUpdate()) {
+          // if srcValue is not a delete, it is a duplicate.
+          return !srcValue.isDelete();
+        } else if (value.isUndelete()) {
+          if (srcValue.isPut()) {
+            throw new IllegalStateException(
+                "An Undelete[" + value + "] and a Put[" + srcValue + "] can't be at the same lifeVersion at store "
+                    + dataDir);
+          }
+          // value is a UNDELETE without a TTL update or a DELETE
+          return !srcValue.isDelete() && !srcValue.isTtlUpdate();
+        } else {
+          if (srcValue.isUndelete()) {
+            throw new IllegalStateException(
+                "A Put[" + value + "] and an Undelete[" + srcValue + "] can't be at the same lifeVersion at store "
+                    + dataDir);
+          }
+          // value is a PUT without a TTL update or a DELETE
+          return !srcValue.isDelete() && !srcValue.isTtlUpdate();
```

Comment:
same here, this can be simplified (as long as comment is here to help us understand the only possible duplicate case is PUT)